uber/models.py

```diff
@@ -533,21 +533,29 @@ def update_badge(self, attendee, old_badge_type, old_badge_num):
             """
             from uber.badge_funcs import needs_badge_num
             old_badge_num = int(old_badge_num or 0) or None
+            was_dupe_num = self.query(Attendee).filter(Attendee.badge_type == old_badge_type,
+                                                       Attendee.badge_num == old_badge_num,
+                                                       Attendee.id != attendee.id).first()
 
-            if old_badge_type == attendee.badge_type and (not attendee.badge_num or old_badge_num == attendee.badge_num):
+            if not was_dupe_num and old_badge_type == attendee.badge_type and (not attendee.badge_num or old_badge_num == attendee.badge_num):
                 attendee.badge_num = old_badge_num
                 return 'Attendee is already {} with badge {}'.format(c.BADGES[old_badge_type], old_badge_num)
 
             if c.SHIFT_CUSTOM_BADGES:
                 # fill in the gap from the old number, if applicable
                 badge_num_keep = attendee.badge_num
-                if old_badge_num:
+                if old_badge_num and not was_dupe_num:
```

Comment:
This fixes part 2 of #2114 by stopping the badges from shifting to "fill in" if the badge being changed had a duplicate number.